File: sdk/swift/Sources/SeocheonSDK/Types/Responses.swift
import Foundation

/// Returned after submitting an activity.
public struct SubmitActivityResponse: Codable, Sendable {
    public let txHash: String
    public let blockHeight: Int64
    public let windowNumber: Int64
    public let epochNumber: Int64
    public let quotaRemaining: UInt64
}

/// A single activity record.
public struct ActivityItem: Codable, Sendable {
    public let activityHash: String
    public let contentUri: String
    public let blockHeight: Int64
    public let windowNumber: Int64
    public let txHash: String
}

/// Returned when querying activities.
public struct GetActivitiesResponse: Codable, Sendable {
    public let activities: [ActivityItem]
    public let totalCount: UInt64
}

/// Returned when querying activity quota.
public struct GetQuotaResponse: Codable, Sendable {
    public let epochNumber: Int64
    public let quotaTotal: UInt64
    public let quotaUsed: UInt64
    public let quotaRemaining: UInt64
    public let isFeegrant: Bool
    public let feegrantExpiry: Int64?
}

/// Returned when querying epoch information.
public struct EpochInfoResponse: Codable, Sendable {
    public let blockHeight: Int64
    public let epochNumber: Int64
    public let epochStartBlock: Int64
    public let epochEndBlock: Int64
    public let epochProgress: String
    public let windowNumber: Int64
    public let windowStartBlock: Int64
    public let windowEndBlock: Int64
    public let windowProgress: String
    public let blocksUntilNextWindow: Int64
    public let blocksUntilNextEpoch: Int64
}

/// Activity within a single window.
public struct WindowActivity: Codable, Sendable {
    public let windowNumber: Int64
    public var submissionCount: UInt64
    public var hasActivity: Bool
}

/// Returned when querying reward qualification.
public struct QualificationResponse: Codable, Sendable {
    public let epochNumber: Int64
    public let totalWindows: Int64
    public let elapsedWindows: Int64
    public let activeWindows: UInt64
    public let requiredWindows: Int64
    public let isQualified: Bool
    public let remainingNeeded: Int64
    public let canStillQualify: Bool
    public let windowDetail: [WindowActivity]
}

/// Returned when querying node information.
public struct NodeInfoResponse: Codable, Sendable {
    public let nodeId: String
    public let `operator`: String
    public let agentAddress: String
    public let status: String
    public let description: String
    public let website: String
    public let tags: [String]
    public let commissionRate: String
    public let agentShare: String
    public let totalDelegation: String
    public let selfDelegation: String
    public let validatorAddress: String
    public let registeredAt: Int64
}

/// A condensed view of a node.
public struct NodeSummary: Codable, Sendable {
    public let nodeId: String
    public let status: String
    public let tags: [String]
    public let totalDelegation: String
    public let description: String
}

/// Returned from a node search.
public struct NodeSearchResponse: Codable, Sendable {
    public let nodes: [NodeSummary]
    public let totalCount: UInt64
}

/// Returned when querying pending rewards.
public struct PendingRewardsResponse: Codable, Sendable {
    public let delegationReward: String
    public let activityReward: String
    public let totalReward: String
    public let commissionTotal: String
    public let operatorShare: String
    public let agentShare: String
}

/// Returned after withdrawing rewards.
public struct WithdrawRewardsResponse: Codable, Sendable {
    public let txHash: String
    public let withdrawnTotal: String
    public let toOperator: String
    public let toAgent: String
}

/// Returned when querying balance.
public struct BalanceResponse: Codable, Sendable {
    public let address: String
    public let balance: String
    public let balanceKkot: String
}

/// Returned after sending tokens.
public struct SendTokensResponse: Codable, Sendable {
    public let txHash: String
    public let blockHeight: Int64
}

/// Returned when querying block information.
public struct BlockInfoResponse: Codable, Sendable {
    public let blockHeight: Int64
    public let blockTime: String
    public let chainId: String
    public let numTxs: UInt64
}

/// An event emitted by a transaction.
public struct TxEvent: Codable, Sendable {
    public let type: String
    public let attributes: [EventAttribute]
}

/// A key-value pair within a TxEvent.
public struct EventAttribute: Codable, Sendable {
    public let key: String
    public let value: String
}

/// Returned when querying a transaction result.
public struct TxResultResponse: Codable, Sendable {
    public let txHash: String
    public let height: Int64
    public let code: UInt32
    public let gasUsed: UInt64
    public let gasWanted: UInt64
    public let rawLog: String
    public let events: [TxEvent]
}

File: sdk/python/seocheon/types/responses.py
"""Response dataclasses for the Seocheon SDK."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class SubmitActivityResponse:
    """Returned after submitting an activity."""

    tx_hash: str
    block_height: int
    window_number: int
    epoch_number: int
    quota_remaining: int


@dataclass
class ActivityItem:
    """Represents a single activity record."""

    activity_hash: str
    content_uri: str
    block_height: int
    window_number: int
    tx_hash: str = ""


@dataclass
class GetActivitiesResponse:
    """Returned when querying activities."""

    activities: list[ActivityItem]
    total_count: int


@dataclass
class GetQuotaResponse:
    """Returned when querying activity quota."""

    epoch_number: int
    quota_total: int
    quota_used: int
    quota_remaining: int
    is_feegrant: bool
    feegrant_expiry: int | None = None


@dataclass
class EpochInfoResponse:
    """Returned when querying epoch information."""

    block_height: int
    epoch_number: int
    epoch_start_block: int
    epoch_end_block: int
    epoch_progress: str
    window_number: int
    window_start_block: int
    window_end_block: int
    window_progress: str
    blocks_until_next_window: int
    blocks_until_next_epoch: int


@dataclass
class WindowActivity:
    """Represents activity within a single window."""

    window_number: int
    submission_count: int
    has_activity: bool


@dataclass
class QualificationResponse:
    """Returned when querying reward qualification."""

    epoch_number: int
    total_windows: int
    elapsed_windows: int
    active_windows: int
    required_windows: int
    is_qualified: bool
    remaining_needed: int
    can_still_qualify: bool
    window_detail: list[WindowActivity] = field(default_factory=list)


@dataclass
class NodeInfoResponse:
    """Returned when querying node information."""

    node_id: str
    operator: str
    agent_address: str
    status: str
    description: str
    website: str
    tags: list[str]
    commission_rate: str
    agent_share: str
    total_delegation: str
    self_delegation: str
    validator_address: str
    registered_at: int


@dataclass
class NodeSummary:
    """A condensed view of a node."""

    node_id: str
    status: str
    tags: list[str]
    total_delegation: str
    description: str


@dataclass
class NodeSearchResponse:
    """Returned from a node search."""

    nodes: list[NodeSummary]
    total_count: int


@dataclass
class PendingRewardsResponse:
    """Returned when querying pending rewards."""

    delegation_reward: str
    activity_reward: str
    total_reward: str
    commission_total: str
    operator_share: str
    agent_share: str


@dataclass
class WithdrawRewardsResponse:
    """Returned after withdrawing rewards."""

    tx_hash: str
    withdrawn_total: str
    to_operator: str
    to_agent: str


@dataclass
class BalanceResponse:
    """Returned when querying balance."""

    address: str
    balance: str
    balance_kkot: str


@dataclass
class SendTokensResponse:
    """Returned after sending tokens."""

    tx_hash: str
    block_height: int


@dataclass
class BlockInfoResponse:
    """Returned when querying block information."""

    block_height: int
    block_time: str
    chain_id: str
    num_txs: int


@dataclass
class EventAttribute:
    """A key-value pair within a TxEvent."""

    key: str
    value: str


@dataclass
class TxEvent:
    """Represents an event emitted by a transaction."""

    type: str
    attributes: list[EventAttribute]


@dataclass
class TxResultResponse:
    """Returned when querying a transaction result."""

    tx_hash: str
    height: int
    code: int
    gas_used: int
    gas_wanted: int
    raw_log: str
    events: list[TxEvent]
